[core] Nullpointer while creating cache File

PMD Version: 5.8.1

I use ant to create a XML PMD Report and want to use cacheLocation.
If I use this the cache file is created but then terminates with following exception.

```
  [pmd:pmd] Sep 07, 2017 3:06:50 PM net.sourceforge.pmd.cache.FileAnalysisCache loadFromFile
  [pmd:pmd] WARNUNG: Cache file <path>\pmd\cachefile is malformed, will not be used for current analysis
  [pmd:pmd] java.lang.NullPointerException
  [pmd:pmd] 	at java.io.DataOutputStream.writeUTF(DataOutputStream.java:347)
  [pmd:pmd] 	at java.io.DataOutputStream.writeUTF(DataOutputStream.java:323)
  [pmd:pmd] 	at net.sourceforge.pmd.cache.CachedRuleViolation.storeToStream(CachedRuleViolation.java:156)
  [pmd:pmd] 	at net.sourceforge.pmd.cache.FileAnalysisCache.persist(FileAnalysisCache.java:111)
  [pmd:pmd] 	at net.sourceforge.pmd.PMD.processFiles(PMD.java:414)
  [pmd:pmd] 	at net.sourceforge.pmd.ant.internal.PMDTaskImpl.doTask(PMDTaskImpl.java:188)
  [pmd:pmd] 	at net.sourceforge.pmd.ant.internal.PMDTaskImpl.execute(PMDTaskImpl.java:269)
  [pmd:pmd] 	at net.sourceforge.pmd.ant.PMDTask.execute(PMDTask.java:49)
  [pmd:pmd] 	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
  [pmd:pmd] 	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
  [pmd:pmd] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  [pmd:pmd] 	at java.lang.reflect.Method.invoke(Method.java:606)
  [pmd:pmd] 	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
  [pmd:pmd] 	at org.apache.tools.ant.Task.perform(Task.java:348)
  [pmd:pmd] 	at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:68)
  [pmd:pmd] 	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
  [pmd:pmd] 	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
  [pmd:pmd] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  [pmd:pmd] 	at java.lang.reflect.Method.invoke(Method.java:606)
  [pmd:pmd] 	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
  [pmd:pmd] 	at org.apache.tools.ant.Task.perform(Task.java:348)
  [pmd:pmd] 	at org.apache.tools.ant.taskdefs.MacroInstance.execute(MacroInstance.java:396)
  [pmd:pmd] 	at sun.reflect.GeneratedMethodAccessor68.invoke(Unknown Source)
  [pmd:pmd] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  [pmd:pmd] 	at java.lang.reflect.Method.invoke(Method.java:606)
  [pmd:pmd] 	at net.sf.antcontrib.logic.For$Reflector.call(Unknown Source)
  [pmd:pmd] 	at net.sf.antcontrib.logic.For.doSequentialIteration(Unknown Source)
  [pmd:pmd] 	at net.sf.antcontrib.logic.For.doTheTasks(Unknown Source)
  [pmd:pmd] 	at net.sf.antcontrib.logic.For.execute(Unknown Source)
  [pmd:pmd] 	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
  [pmd:pmd] 	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
  [pmd:pmd] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  [pmd:pmd] 	at java.lang.reflect.Method.invoke(Method.java:606)
  [pmd:pmd] 	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
  [pmd:pmd] 	at org.apache.tools.ant.Task.perform(Task.java:348)
  [pmd:pmd] 	at org.apache.tools.ant.Target.execute(Target.java:435)
  [pmd:pmd] 	at org.apache.tools.ant.Target.performTasks(Target.java:456)
  [pmd:pmd] 	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
  [pmd:pmd] 	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
  [pmd:pmd] 	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
  [pmd:pmd] 	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
  [pmd:pmd] 	at org.apache.tools.ant.Main.runBuild(Main.java:851)
  [pmd:pmd] 	at org.apache.tools.ant.Main.startAnt(Main.java:235)
  [pmd:pmd] 	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:280)
  [pmd:pmd] 	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:109)

BUILD FAILED
<path>\build.xml:2324: The following error occurred while executing this line:
<path>\build.xml:2334: java.lang.NullPointerException
	at net.sourceforge.pmd.util.StringUtil.appendXmlEscaped(StringUtil.java:257)
	at net.sourceforge.pmd.renderers.XMLRenderer.end(XMLRenderer.java:132)
	at net.sourceforge.pmd.ant.Formatter.end(Formatter.java:117)
	at net.sourceforge.pmd.ant.internal.PMDTaskImpl.doTask(PMDTaskImpl.java:198)
	at net.sourceforge.pmd.ant.internal.PMDTaskImpl.execute(PMDTaskImpl.java:269)
	at net.sourceforge.pmd.ant.PMDTask.execute(PMDTask.java:49)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:68)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.taskdefs.MacroInstance.execute(MacroInstance.java:396)
	at sun.reflect.GeneratedMethodAccessor68.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at net.sf.antcontrib.logic.For$Reflector.call(Unknown Source)
	at net.sf.antcontrib.logic.For.doSequentialIteration(Unknown Source)
	at net.sf.antcontrib.logic.For.doTheTasks(Unknown Source)
	at net.sf.antcontrib.logic.For.execute(Unknown Source)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1393)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1364)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1248)
	at org.apache.tools.ant.Main.runBuild(Main.java:851)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:280)
	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:109)
```


**Running PMD through:** *Ant*
**Ant call**
```
<pmd:pmd shortFilenames="true">
    <sourceLanguage name="java" version="1.7"/>
    <classpath>
        <pathelement location="src/main/java" />
    </classpath>
    <auxclasspath>
        <path refid="pmd.auxpath" />
     </auxclasspath>
     <ruleset>pmd_ruleset.xml</ruleset>
      <formatter type="xml" toFile="pmd-report.xml" />
      <fileset dir="${module.src.cqm.path.entryc.dir}" includes="**/*.java" />
</pmd:pmd>
```

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -152,6 +152,8 @@ All existing rules have been updated to reflect these changes. If you have custo
 
 ### Fixed Issues
 
+*   all
+    *   [#600](https://github.com/pmd/pmd/issues/600): \[core] Nullpointer while creating cache File
 *   apex
     *   [#488](https://github.com/pmd/pmd/pull/488): \[apex] Use Apex lexer for CPD
     *   [#489](https://github.com/pmd/pmd/pull/489): \[apex] Update Apex compiler

--- a/pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java
@@ -145,15 +145,19 @@ public final class CachedRuleViolation implements RuleViolation {
      */
     /* package */ static void storeToStream(final DataOutputStream stream,
             final RuleViolation violation) throws IOException {
-        stream.writeUTF(violation.getDescription());
-        stream.writeUTF(violation.getRule().getRuleClass());
+        stream.writeUTF(getValueOrEmpty(violation.getDescription()));
+        stream.writeUTF(getValueOrEmpty(violation.getRule().getRuleClass()));
         stream.writeInt(violation.getBeginLine());
         stream.writeInt(violation.getBeginColumn());
         stream.writeInt(violation.getEndLine());
         stream.writeInt(violation.getEndColumn());
-        stream.writeUTF(violation.getPackageName());
-        stream.writeUTF(violation.getClassName());
-        stream.writeUTF(violation.getMethodName());
-        stream.writeUTF(violation.getVariableName());
+        stream.writeUTF(getValueOrEmpty(violation.getPackageName()));
+        stream.writeUTF(getValueOrEmpty(violation.getClassName()));
+        stream.writeUTF(getValueOrEmpty(violation.getMethodName()));
+        stream.writeUTF(getValueOrEmpty(violation.getVariableName()));
+    }
+
+    private static String getValueOrEmpty(final String value) {
+        return value == null ? "" : value;
     }
 }

--- a/pmd-core/src/test/java/net/sourceforge/pmd/cache/FileAnalysisCacheTest.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/cache/FileAnalysisCacheTest.java
@@ -91,7 +91,7 @@ public class FileAnalysisCacheTest {
         final FileAnalysisCache cache = new FileAnalysisCache(newCacheFile);
         cache.isUpToDate(sourceFile);
 
-        final RuleViolation rv = mock(RuleViolation.class, Mockito.RETURNS_SMART_NULLS);
+        final RuleViolation rv = mock(RuleViolation.class);
         when(rv.getFilename()).thenReturn(sourceFile.getPath());
         final net.sourceforge.pmd.Rule rule = mock(net.sourceforge.pmd.Rule.class, Mockito.RETURNS_SMART_NULLS);
         when(rv.getRule()).thenReturn(rule);

